### PR TITLE
Payment line items

### PR DESCRIPTION
**Pull request by torotil**
Depends on #4

### CODE REVIEW

File: docs/source/events/index.rst [hunk 1]
```diff
@@ -54,6 +54,7 @@ All events are encoded as JSON. On top of the JSON built-in types the events can
 - date: ISO 8601 encoded dates either with an explicit offset or in UTC
 - interval: ISO 8601 time interval
 - money: Money value encoded as string of the floating point value and the ISO 4217 code separated by a space
+- float: A floating point number encoded as string.
```
Comment (edjw) [inline docs/source/events/index.rst]: I think give an example of floating point number just in case someone isn't familiar or maybe the programming language they know best uses a different name for this or doesn't have a name for this, for example:

```rst
- float: A floating point number encoded as string such as 1.0 or 2.0
```

I think we don't support donations that are not whole numbers. If we actually do support donations of £2.50 then use .5 in one of the examples

File: docs/source/events/index.rst [hunk 2]
```diff
@@ -54,6 +54,7 @@ All events are encoded as JSON. On top of the JSON built-in types the events can
 - date: ISO 8601 encoded dates either with an explicit offset or in UTC
 - interval: ISO 8601 time interval
 - money: Money value encoded as string of the floating point value and the ISO 4217 code separated by a space
```
Comment (edjw) [inline docs/source/events/index.rst]: Give an example of what this might look like

File: docs/source/events/index.rst [hunk 3]
```diff
@@ -54,6 +54,7 @@ All events are encoded as JSON. On top of the JSON built-in types the events can
 - date: ISO 8601 encoded dates either with an explicit offset or in UTC
```
Comment (edjw) [inline docs/source/events/index.rst]: Why might it be a) with an explicit offset, or b) in UTC? Any reason why it might vary? I think people will want to know which to expect.

Could also give an example so people don't have to look up ISO 8601

File: docs/source/events/index.rst [hunk 4]
```diff
@@ -54,6 +54,7 @@ All events are encoded as JSON. On top of the JSON built-in types the events can
...
 - interval: ISO 8601 time interval
```
Comment (edjw) [inline docs/source/events/index.rst]: Give an example so people don't have to look this up

File: docs/source/events/examples/payment-status-change.yaml [hunk 5]
```diff
@@ -17,5 +17,17 @@ currency_code: EUR
...
+- name: one-off
+  amount: 10
+  quantity: 9
+  tax_rate: 0
+  recurrence_interval: null
+- name: monthly-item
+  amount: 10
+  quantity: 1
+  tax_rate: 0
+  recurrence_interval: P1M
```
Comment (edjw) [inline docs/source/events/examples/payment-status-change.yaml:31]: What can the recurrence intervals be? What does P1M mean? What else could it be?

File: docs/source/events/examples/payment-status-change.yaml [hunk 6]
```diff
@@ -17,5 +17,17 @@ currency_code: EUR
 # Payment method specific data (stripe intent ID in this case)
 payment_data:
   transaction_id: pi_3NOKIODvbS6ezf4w21zIR6Ft
+# Line items:
+line_items:
```
Comment (edjw) [inline docs/source/events/examples/payment-status-change.yaml:21]: Should currency be here?

File: docs/source/events/examples/payment-status-change.yaml [hunk 7]
```diff
@@ -17,5 +17,17 @@ currency_code: EUR
...
+- name: one-off
+  amount: 10
```
Comment (edjw) [inline docs/source/events/examples/payment-status-change.yaml:23]: This is an integer not a float? If I'm right that we don't support donation of £2.50, then we could say that so developers know they can treat this as an integer

File: docs/source/events/examples/payment-status-change.yaml [hunk 8]
```diff
@@ -17,5 +17,17 @@ currency_code: EUR
...
+  quantity: 9
```
Comment (edjw) [inline docs/source/events/examples/payment-status-change.yaml:24]: What does `quantity: 9` mean? 9 one-off donations? If so, seems unlikely that this would happen so it's confusing. We could say `1`. 9 would be if someone bought 9 t-shirts through a simple shop on Impact Stack right?

File: docs/source/events/examples/payment-status-change.yaml [hunk 9]
```diff
@@ -17,5 +17,17 @@ currency_code: EUR
...
+  tax_rate: 0
+  recurrence_interval: null
+- name: monthly-item
```
Comment (edjw) [inline docs/source/events/examples/payment-status-change.yaml:27]: why is it `monthly-item` with `-item` on the end? One-off is just `one-off`. Do we support annual? If so, say what the possible values could be

File: docs/source/events/index.rst (hunk 1 above)
Comment (torotil) [reply docs/source/events/index.rst]: What if the API does support it, but the forms don’t? For IST 1.0 they don’t for IST 2.0 they might.

File: docs/source/events/payment-status-change.rst [hunk 10]
```diff
@@ -6,6 +6,30 @@ This event means that the status of a payment has changed. To reduce noise, we o
 For each payment method, there are typical status sequences that the payment goes through. The exact meaning of the statuses and content of the ``payment_data`` also depends on the payment method.
 
 
+Line items
+----------
+
+.. versionadded:: 1.2.0
```
Comment (edjw) [inline docs/source/events/payment-status-change.rst:30]: I assume `versionadded` is in the correct case here but maybe not?

File: docs/source/events/payment-status-change.rst [hunk 11]
```diff
@@ -6,6 +6,30 @@ This event means that the status of a payment has changed. To reduce noise, we o
...
+
+Each status update contains the full list of payment ``line_items``. Each line item has the following data:
+
+- ``name``: A unique name for this line item
+- ``amount`` (float): The price of each unit
```
Comment (edjw) [inline docs/source/events/payment-status-change.rst:35]: Oh this is a float not an integer! The example higher up should be 10.0 then right?

File: docs/source/events/payment-status-change.rst (hunk 10 above)
Comment (torotil) [reply docs/source/events/payment-status-change.rst:30]: Yep, that has a syntactic meaning in RST/Sphinx and gets expanded into an english sentence like: Added in version 1.2.0.

File: docs/source/events/payment-status-change.rst [hunk 12]
```diff
@@ -6,6 +6,30 @@ This event means that the status of a payment has changed. To reduce noise, we o
...
+- ``quantity`` (float): The number of items
+- ``tax_rate`` (float): The applicable VAT (or equivalent) tax rate
+- ``recurrence_interval`` (interval): For recurring payments this specifies the interval (e.g. ``P1M`` for monthly, ``P1Y`` for yearly). It is set to ``null`` for one-off payments.
```
Comment (edjw) [inline docs/source/events/payment-status-change.rst:38]: I suggested examples here earlier. This looks fine. Not sure how it'll look on the webpage from these docs but should be ok?

File: docs/source/events/examples/payment-status-change.yaml (hunk 9 above)
Comment (torotil) [reply docs/source/events/examples/payment-status-change.yaml:27]: I consciously chose to *not* make it consistent because those are names configured by editors in the UI. For the tests (and this examples) they give a way to document it additionally. I still want to avoid anyone expecting the interval being encoded into the `name`.

Comment (edjw) [reply docs/source/events/examples/payment-status-change.yaml:27]: Ok I think that's fine. In that case, we should say that the value is configured in the UI then and reflects whatever it says there

File: docs/source/events/payment-status-change.rst [hunk 13]
```diff
@@ -6,6 +6,30 @@ This event means that the status of a payment has changed. To reduce noise, we o
...
+The total amount of the payment is the sum over all line items:
+
+.. math::
+
+  \sum_{i} amount_i \cdot quantity_i \cdot (1 + tax\_rate_i)
+
+For most of our donation forms there will be only one line item with ``quantity=1`` and ``tax_rate=0``.
+
+The line items are summarized into one :ref:`donation_event` per ``recurrence_interval``. These are perhaps easier to interpret for most use cases.
+
```
Comment (katzmo) [inline docs/source/events/payment-status-change.rst]: Although this accurately describes what happens in the code, it's maybe a bit hard to follow? I’d need something that states more clearly that there is this other kind of event I should be interested in instead, like "The line items  for each ``recurrence_interval`` are summarized in the respective :ref:`donation_event`  – for most use cases, these are easier to interpret." (just a suggestion)

I think I'd also put the payment statuses first – they seem more interesting for this kind of event? :thinking:

File: docs/source/events/examples/payment-status-change.yaml (hunk 8 above)
Comment (katzmo) [reply docs/source/events/examples/payment-status-change.yaml:24]: 9 tickets to an event, 9 christmas presents,… :wink:

File: docs/source/events/examples/payment-status-change.yaml (hunk 5 above)
Comment (torotil) [reply docs/source/events/examples/payment-status-change.yaml:31]: Again this is an example embedded in the main article about this type of event (`payment-status-change.rst`).

File: docs/source/events/examples/payment-status-change.yaml (hunk 6 above)
Comment (torotil) [reply docs/source/events/examples/payment-status-change.yaml:21]: No, IST 1.0 only allows one currency for all the line items.

File: docs/source/events/examples/payment-status-change.yaml (hunk 7 above)
Comment (torotil) [reply docs/source/events/examples/payment-status-change.yaml:23]: At the moment our forms don’t allow non-integer values. The value here is a “float encoded as string“ (see `index.rst`).

If we specify here that this can only be integers then introducing floats in IST 2.0 forms would be a breaking change.

File: docs/source/events/examples/payment-status-change.yaml (hunk 8 above)
Comment (torotil) [reply docs/source/events/examples/payment-status-change.yaml:24]: The main documentation for payment_status_change events (`payment-status-change.rst`) explains that this is usually `quantity=1` … and that you only need to care about these events for advanced use cases.

File: docs/source/events/examples/payment-status-change.yaml (hunk 9 above)
Comment (torotil) [reply docs/source/events/examples/payment-status-change.yaml:27]: Ok, I’ve added half a sentence to where the property is documented. This is what it looks like now.

image

File: docs/source/events/payment-status-change.rst (hunk 13 above)
Comment (torotil) [reply docs/source/events/payment-status-change.rst]: Ok, I’ve implement both of the suggestions.

File: docs/source/events/payment-status-change.rst (hunk 12 above)
Comment (torotil) [reply docs/source/events/payment-status-change.rst:38]: I’ve posted a screenshot above.

File: docs/source/events/payment-status-change.rst (hunk 11 above)
Comment (torotil) [reply docs/source/events/payment-status-change.rst:35]: See the updated documentation of the float type.

File: docs/source/events/index.rst (hunk 4 above)
Comment (torotil) [reply docs/source/events/index.rst]: I have added examples.

File: docs/source/events/index.rst (hunk 3 above)
Comment (torotil) [reply docs/source/events/index.rst]: I have added examples.

Treating a missing offset as ±00:00 shouldn’t be hard to implement. Ensuring either a) or b) in IST 1.0 and IST 2.0 is around 1d of work.

File: docs/source/events/index.rst (hunk 2 above)
Comment (torotil) [reply docs/source/events/index.rst]: I’ve moved this after the float, and added examples.

File: docs/source/events/index.rst (hunk 1 above)
Comment (torotil) [reply docs/source/events/index.rst]: I’ve added examples for it.

File: docs/source/events/payment-status-change.rst (hunk 13 above)
Comment (katzmo) [reply docs/source/events/payment-status-change.rst]: Thanks! The first sentence is missing a verb now :wink:
